File: lib/models/payslip.dart
class Payslip {
  final String id;
  final String employeeId;
  final String employeeName;
  final String month;
  final int year;
  final double basicSalary;
  final double allowances;
  final double deductions;
  final double netSalary;
  final DateTime generatedDate;
  final Map<String, double> allowanceBreakdown;
  final Map<String, double> deductionBreakdown;
  final double overtime;
  final double bonus;
  final String status;
  final String currency;
  final DateTime? payPeriodStart;
  final DateTime? payPeriodEnd;
  final String? generatedBy;
  final String? finalizedBy;
  final DateTime? finalizedDate;
  final String? notes;
 
   Payslip({
    required this.id,
    required this.employeeId,
    required this.employeeName,
    required this.month,
    required this.year,
    required this.basicSalary,
    required this.allowances,
    required this.deductions,
    required this.netSalary,
    required this.generatedDate,
    this.allowanceBreakdown = const {},
    this.deductionBreakdown = const {},
    this.overtime = 0,
    this.bonus = 0,
    this.status = 'Draft',
    this.currency = 'USD',
    this.payPeriodStart,
    this.payPeriodEnd,
    this.generatedBy,
    this.finalizedBy,
    this.finalizedDate,
    this.notes,
  });
 
   double get grossSalary => basicSalary + allowances + overtime + bonus;
 
   Map<String, dynamic> toJson() {
     return {
       'id': id,
       'employeeId': employeeId,
       'employeeName': employeeName,
       'month': month,
       'year': year,
       'basicSalary': basicSalary,
       'allowances': allowances,
       'deductions': deductions,
       'netSalary': netSalary,
       'generatedDate': generatedDate.toIso8601String(),
       'allowanceBreakdown': allowanceBreakdown,
       'deductionBreakdown': deductionBreakdown,
       'overtime': overtime,
       'bonus': bonus,
       'status': status,
       'currency': currency,
       'payPeriodStart': payPeriodStart?.toIso8601String(),
       'payPeriodEnd': payPeriodEnd?.toIso8601String(),
       'generatedBy': generatedBy,
       'finalizedBy': finalizedBy,
       'finalizedDate': finalizedDate?.toIso8601String(),
       'notes': notes,
     };
   }
 
   factory Payslip.fromJson(Map<String, dynamic> json) {
    DateTime? _parseDate(dynamic value) {
      if (value == null) return null;
      if (value is DateTime) return value;
      return DateTime.tryParse(value.toString());
    }

    Map<String, double> _parseBreakdown(dynamic value) {
      if (value is Map) {
        return value.map((key, dynamic val) => MapEntry(
              key.toString(),
              (val is num ? val.toDouble() : double.tryParse(val.toString()) ?? 0),
            ));
      }
      return {};
    }

    return Payslip(
      id: json['id'] as String,
      employeeId: json['employeeId'] as String,
      employeeName: json['employeeName'] as String,
      month: json['month'] as String,
      year: json['year'] is int ? json['year'] as int : int.parse(json['year'].toString()),
      basicSalary: (json['basicSalary'] as num).toDouble(),
      allowances: (json['allowances'] as num).toDouble(),
      deductions: (json['deductions'] as num).toDouble(),
      netSalary: (json['netSalary'] as num).toDouble(),
      generatedDate: _parseDate(json['generatedDate']) ?? DateTime.now(),
      allowanceBreakdown: _parseBreakdown(json['allowanceBreakdown']),
      deductionBreakdown: _parseBreakdown(json['deductionBreakdown']),
      overtime: json['overtime'] != null ? (json['overtime'] as num).toDouble() : 0,
      bonus: json['bonus'] != null ? (json['bonus'] as num).toDouble() : 0,
      status: json['status']?.toString() ?? 'Draft',
      currency: json['currency']?.toString() ?? 'USD',
      payPeriodStart: _parseDate(json['payPeriodStart']),
      payPeriodEnd: _parseDate(json['payPeriodEnd']),
      generatedBy: json['generatedBy']?.toString(),
      finalizedBy: json['finalizedBy']?.toString(),
      finalizedDate: _parseDate(json['finalizedAt']),
      notes: json['notes']?.toString(),
    );
  }
}

File: backend/src/models/Payslip.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../database/connection');

const Payslip = sequelize.define('Payslip', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  organization_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'organizations',
      key: 'id',
    },
    onDelete: 'CASCADE',
  },
  employee_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'employees',
      key: 'id',
    },
    onDelete: 'CASCADE',
  },
  month: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  year: {
    type: DataTypes.INTEGER,
    allowNull: false,
  },
  pay_period_start: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  pay_period_end: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  basic_salary: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
  },
  allowances: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
  },
  allowance_breakdown: {
    type: DataTypes.JSONB,
    allowNull: false,
    defaultValue: {},
  },
  deductions: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
  },
  deduction_breakdown: {
    type: DataTypes.JSONB,
    allowNull: false,
    defaultValue: {},
  },
  overtime: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
  },
  bonus: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
    defaultValue: 0,
  },
  gross_salary: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
  },
  net_salary: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
  },
  currency: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: 'USD',
  },
  status: {
    type: DataTypes.ENUM('draft', 'processed', 'paid'),
    allowNull: false,
    defaultValue: 'draft',
  },
  notes: {
    type: DataTypes.TEXT,
    allowNull: true,
  },
  generated_by: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'users',
      key: 'id',
    },
    onDelete: 'SET NULL',
  },
  generated_at: {
    type: DataTypes.DATE,
    allowNull: false,
    defaultValue: DataTypes.NOW,
  },
  finalized_by: {
    type: DataTypes.UUID,
    allowNull: true,
    references: {
      model: 'users',
      key: 'id',
    },
    onDelete: 'SET NULL',
  },
  finalized_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
}, {
  tableName: 'payslips',
  indexes: [
    {
      fields: ['organization_id', 'employee_id'],
      name: 'payslips_org_employee',
    },
    {
      fields: ['employee_id', 'year', 'month'],
      unique: true,
      name: 'unique_payslip_per_month',
    },
    {
      fields: ['status'],
    },
  ],
});

module.exports = Payslip;
